### DISCUSSION TITLE
Can I safely use tokio's task_local on a per-request basis?

### DISCUSSION THREAD

**Discussion by edgarogh** [tokio-rs/axum#2631] [Q&A] [answered]
### Summary

Hello, I'm wondering if axum handles requests within tokio tasks or not? If yes, is it safe to rely on this property to have a task_local value set by a middleware? (and only accessible for the current request)

If that's not the case, I know how to implement my own custom middleware future that sets and unsets a thread-local value respectively before and after polling. I simply prefer to use existing framework features, if any.

_I am 100% aware that using thread_local/task_local stuff is very unidiomatic and could a big code smell, but in my specific case, I'm pretty sure that my current situation is way worse in terms of simplicity and maintainability._

### axum version

0.6.18

**Comment by jplatte** [comment, marked as answer]
Yes, you can use a task local to pass information about the request to inner middlewares, handlers and `IntoResponse` impls. Though for all but the last one, you can also use request [extensions](https://docs.rs/http/1.0.0/http/request/struct.Request.html#method.extensions).

If is however also possible for inner middleware to do parts of its work (including calling the inner service) on different tasks though, making task locals unavailable. This is rare and something I would expect the middleware's docs to point out, though.

**Reply by edgarogh** [reply]
That's perfect, thank you! Extensions are unfortunately out of the question here, that's why I want to use global state (despite its disadvantages).

Is it mentioned in the docs somewhere, that `task_local!` is safe to use in handlers/middlewares/IntoResponse? I think it can be worth documenting, not without mentioning that other means of passing information around are still much more preferable. I'd happily submit a PR (I'm not sure on which doc(s) item, though).

**Reply by jplatte** [reply]
It's not currently mentioned in the docs. A PR would be appreciated, I think the "Passing state from middleware to handlers" section would be a good place. And I think we should have an example like

```rust
axum::middleware::from_fn(|..., next| async move {
    task_local! { ... };
    // Make MY_TASK_LOCAL available to inner middleware,
    // handlers, `FromRequest` and `IntoResponse` impls
    MY_TASK_LOCAL.scope(async move {
        next.run().await
    })
})
```
